🚧 [new module] Environment variables editor

stefansjfw opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request

Co-authored very much by @niels9001 

Environment Variables app offers a nice and easy way to manage environment variables. Also, you can create a profile (set of environment variables) to apply multiple variables at once. Applying the profile adds variables to User env variables. On profile apply, If there is an existing variable with the same name in User variables, backup <var name>_PowerToys_<profile name> variable is created to be able to revert it on profile un-apply (and prevent possible issues causing lost variable values).

The applied variables pane on the right shows the status of applied variables, taking in order which (duplicated or not) variables take precedence, i.e. Profile -> User -> System.

TODO:
[ ] Icon
[ ] Settings and OOBE screenshots
[ ] Docs PR

![image](https://github.com/microsoft/PowerToys/assets/57057282/20ac1e0a-0e5e-4e9c-82e5-b8edbeada091)

![image](https://github.com/microsoft/PowerToys/assets/57057282/088769e7-f5a3-43c5-87fd-c7c19f329e38)

![image](https://github.com/microsoft/PowerToys/assets/57057282/436b6744-14d4-45e4-b468-5d4ec3931e66)

![image](https://github.com/microsoft/PowerToys/assets/57057282/937cf656-9298-49b6-8f1c-ba52f147920d)

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [x] **Closes:** #16424
- [x] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [x] **Tests:** Added/updated and all pass
- [x] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated
- [x] **New binaries:** Added on the required places
   - [x] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [x] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [x] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [x] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

stefansjfw committed:
Init EnvironmentVariables UI project

stefansjfw committed:
Init EnvironmentVariables UI project

stefansjfw committed:
Models
TitleBar
MainPage init
Icon

stefansjfw committed:
Models
TitleBar
MainPage init
Icon

stefansjfw committed:
User and system variables

stefansjfw committed:
User and system variables

stefansjfw committed:
Profiles init

stefansjfw committed:
Profiles init

niels9001 committed:
XAML cleanup

niels9001 committed:
XAML cleanup

niels9001 committed:
Missing ItemTemplate

niels9001 committed:
Missing ItemTemplate

stefansjfw committed:
EditDialog

stefansjfw committed:
EditDialog

stefansjfw committed:
ModuleInterface

stefansjfw committed:
ModuleInterface

stefansjfw committed:
Signing and processes lists

stefansjfw committed:
Signing and processes lists

stefansjfw committed:
Installer

stefansjfw committed:
Installer

stefansjfw committed:
spellcheck

stefansjfw committed:
spellcheck

stefansjfw committed:
Fix ARM64 build and consolidate packages

stefansjfw committed:
Fix ARM64 build and consolidate packages

stefansjfw committed:
spellcheck2

stefansjfw committed:
spellcheck2

stefansjfw committed:
Fix installer

stefansjfw committed:
Fix installer

stefansjfw committed:
Single instance. C# telemetry. Wait on PT pid

stefansjfw committed:
Single instance. C# telemetry. Wait on PT pid

stefansjfw committed:
ElevationHelper

stefansjfw committed:
ElevationHelper

stefansjfw committed:
Add profile wip

stefansjfw committed:
Add profile wip

davidegiacometti committed:
show run as administrator in title (#28516)

davidegiacometti committed:
Environment Variables added to Run plugin (#28466)

stefansjfw committed:
Merge branch 'stefan/env_var_new' of https://github.com/microsoft/PowerToys into stefan/env_var_new

stefansjfw committed:
Merge remote-tracking branch 'origin/main' into stefan/env_var_new

niels9001 committed:
UI tweaks

niels9001 committed:
Remove style

stefansjfw committed:
Add profile - init working

stefansjfw committed:
Applied variables

stefansjfw committed:
Read/Write profiles

niels9001 committed:
Fixes

stefansjfw committed:
Add separator and fix loading profiles

stefansjfw committed:
Only allow to edit System vars if running elevated

stefansjfw committed:
Add tmp progress ring to show applying changes progress
Ignore not needed json fields

stefansjfw committed:
Remove variable and profile logic

stefansjfw committed:
Merge remote-tracking branch 'origin/main' into stefan/env_var_new

stefansjfw committed:
Do not read data async
Update System and User variables on change

stefansjfw committed:
Add isCorrectlyApplied()

stefansjfw committed:
Merge remote-tracking branch 'origin/main' into stefan/env_var_new

stefansjfw committed:
Sort variables in Applied variables

stefansjfw committed:
WIP WndProc

stefansjfw committed:
spellcheck

stefansjfw committed:
Revert "WIP WndProc"

This reverts commit 0c0b6c67de1f8ce2c18265a3fa7e044942983eba.

stefansjfw committed:
WHY CRASH???

niels9001 committed:
UI tweaks

stefansjfw committed:
WIP modified state warning

stefansjfw committed:
Add cancel button in dialogs

stefansjfw committed:
Add buttons validations

stefansjfw committed:
Set variables - fire and forget notify

stefansjfw committed:
Revert "Revert "WIP WndProc""

This reverts commit 1b2306eeb73b3a49ad0e6f540d1037f078222884.

stefansjfw committed:
Listen to WM_SETTINGSCHANGED
Add Infobar reload button

stefansjfw committed:
spellcheck

stefansjfw committed:
spellcheck again

stefansjfw committed:
Merge remote-tracking branch 'origin/main' into stefan/env_var_new

stefansjfw committed:
Fix build

stefansjfw committed:
InfoBar runAsAdmin visibility

stefansjfw committed:
Fix comment

davidegiacometti reviewed (commented):
Although I haven't looked at the code yet I have tested it on Windows 10.
UI/UX is very nice compared to the Windows native Environment Variables tool.

I have a few nits and improvements:
- You can't create a new profile with only existing variables: the **Add** button stay disabled.
- You can't add/remove variables from an existing profiles.
- You can't add/remove variables directly from user/system variables.
- User/System variables aren't sorted alphabetically.
- Let's add windows minimum width/height

![image](https://github.com/microsoft/PowerToys/assets/25966642/2a3cb9a3-7d47-44eb-9447-cd60e57e5cc8)

stefansjfw commented:
> Although I haven't looked at the code yet I have tested it on Windows 10. UI/UX is very nice compared to the Windows native Environment Variables tool.
> 
> I have a few nits and improvements:
> 

Thanks for giving it at try :) 

> * You can't create a new profile with only existing variables: the **Add** button stay disabled.
Good catch. I'll fix it.

> * You can't add/remove variables from an existing profiles.
You can remove with right-click context menu (although it clashes with Select all for variable value). Adding is WIP.

> * You can't add/remove variables directly from user/system variables.
You can remove with right-click context menu (although it clashes with Select all for variable value). Adding is WIP.

> * User/System variables aren't sorted alphabetically.
I'll add this.

> * Let's add windows minimum width/height
👍

davidegiacometti commented:
Oh.... The inline edit button made me miss the context menu!
I guess a confirm dialog for delete would be nice.

htcfreek commented:
@stefansjfw
Some feedback about the list of "applied variables":
1. I can't see the complete value of long variables like "path". A tool tip would be good here.
2. The path variable preview is wrong. It has to add the user value to the end of the machine value instead of replacing the machine value.
3. The handling of name casing is wrong. If you have a computer variable named "testVar" and a user variable named "TestVar" then the merged variable has to use the name of the machine variable with the value of the user variable (e.g. `testVar=UserValue`). Currently two variables are shown because the dictionary is not case insensitive.

htcfreek commented on src/common/GPOWrapper/GPOWrapper.cpp:
```diff
@@ -144,4 +144,8 @@ namespace winrt::PowerToys::GPOWrapper::implementation
     }
+    GpoRuleConfigured GPOWrapper::GetConfiguredEnvironmentVariablesEnabledValue()
+    {
+        return static_cast<GpoRuleConfigured>(powertoys_gpo::getConfiguredEnvironmentVariablesEnabledValue());
```
If #28731 gets merged the way of implementing the utility policies changes a bit. Please be aware of this change.

htcfreek commented on src/common/utils/gpo.h:
```diff
@@ -305,4 +306,8 @@ namespace powertoys_gpo {
 
+    inline gpo_rule_configured_t getConfiguredEnvironmentVariablesEnabledValue()
+    {
+        return getConfiguredValue(POLICY_CONFIGURE_ENABLED_ENVIRONMENT_VARIABLES);
```
If https://github.com/microsoft/PowerToys/pull/28731 gets merged the way of implementing the utility policies changes a bit. Please be aware of this change.

htcfreek replied:
PR is merged. Please update your new code.

stefansjfw replied:
thanks. I'll update it

stefansjfw replied:
done

stefansjfw committed:
Confirm dialog when deleting variable
Fix add variable button when creating profile

stefansjfw committed:
Edit profile

stefansjfw committed:
Sort variables on Load

stefansjfw committed:
Select existing variables on edit

stefansjfw committed:
Add default variable

stefansjfw committed:
Merge remote-tracking branch 'origin/main' into stefan/env_var_new

stefansjfw committed:
Fix adding existing vars to profile

stefansjfw committed:
update notice.md

stefansjfw commented:
> Although I haven't looked at the code yet I have tested it on Windows 10. UI/UX is very nice compared to the Windows native Environment Variables tool.
> 
> I have a few nits and improvements:
> 
> * You can't create a new profile with only existing variables: the **Add** button stay disabled.
> * You can't add/remove variables from an existing profiles.
> * You can't add/remove variables directly from user/system variables.
> * User/System variables aren't sorted alphabetically.
> * Let's add windows minimum width/height
> 
> ![image](https://user-images.githubusercontent.com/25966642/270099839-2a3cb9a3-7d47-44eb-9447-cd60e57e5cc8.png)

these should be fixed now :) Also confirmation dialog when deleting a variable is added.

stefansjfw commented:
> @stefansjfw
> 
> **Some feedback about the list of applied variables:**
> 
> 1. I can't see the complete value of long variables like "path". A tool tip would be good here.

Tooltips are missing everywhere, will add those.

> 2. The path variable preview is wrong. It has to add the user value to the end of the machine value instead of replacing the machine value.

Fixed.

> 3. The handling of name casing is wrong. If you have a computer variable named "testVar" and a user variable named "TestVar" then the merged variable has to use the name of the machine variable with the value of the user variable (e.g. `testVar=UserValue`). Currently two variables are shown because the dictionary is not case insensitive.
> 

I'll fix this, thanks!

> **Some feedback about the profiles:**
> 
> 1. I wondering if we should support defining computer and user profiles instead of only user profiles. (if not we should add a hint that the variables are applied as user variables.)

Let's see if this will be asked for I guess. 

> 2. The text in the delete confirmation dialog is cut off.
>    ![image](https://user-images.githubusercontent.com/61519853/270122139-032301d9-5b63-4ecf-84aa-c01da5b22609.png)
> 
> **Some feedback about the non-admin mode:**
> 

Will fix it. Thanks.

> 1. I suggest to add a restart as admin button to the UI.

I'll add it.

> 2. If the blue dot in front of the Admin info message should be a info badge then the `i` is missing. Otherwise I suggest to change the symbol to the info symbol.
>    ![image](https://user-images.githubusercontent.com/61519853/270122282-5b8da0e0-e528-43ad-9ec7-4228a6ef31b6.png)

Will be fixed

stefansjfw committed:
Handle PATH properly

htcfreek commented on src/modules/EnvironmentVariables/EnvironmentVariables/ViewModels/MainViewModel.cs:
```diff
@@ -0,0 +1,336 @@
+                var clone = systemPath.Clone();
+                clone.Values += ";" + userPath.Values;
+                variables.Remove(userPath);
+                variables.Insert(variables.IndexOf(systemPath), clone);
```
I think this should either be marked as user scoope or have it's own type as it contains `computer + user` value.

stefansjfw replied:
hm, makes sense to make it obvious. not sure which one to go for though

htcfreek replied:
What about the path symbol. It's a special var and therefore it has special handling.

By the way, it makes a difference in zhe Windows env var ui if the name is `PATH` or `Path`. If it is `PATH` a special list dialogue opens on editing.

stefansjfw replied:
> What about the path symbol. It's a special var and therefore it has special handling.

yeah, that one too. I'll see what to do there (for both path and duplicate vars)

> By the way, it makes a difference in zhe Windows env var ui if the name is PATH or Path. If it is PATH a special list dialogue opens on editing.

Will be added (for v2?). Trying to figure out the best UI for it.

stefansjfw replied:
Some of these maybe E7EE EE57 EF58 F180 F6A9 ? For Path and duplicates? WDYT?

htcfreek replied:
For the path var I would use E8AC (Rename). And for duplicated vars I would use E8C8 (Copy) or E8C9 (Important). The icons for system and user looked good to me.

htcfreek commented:
@stefansjfw
>> 1. I can't see the complete value of long variables like "path". A tool tip would be good here.

> Tooltips are missing everywhere, will add those.

I think a left/right scrollbar in the applied vars list would be a good idea too.

stefansjfw committed:
Merge remote-tracking branch 'origin/main' into stefan/env_var_new

stefansjfw committed:
Add tooltips and fix dialogs text wrapping

stefansjfw committed:
Fix applied values for duplicates
Fix add/eddit variable txt box validation

stefansjfw committed:
Add horizontal scroll bar for applied values

stefansjfw committed:
Fix duplicate variables handling
Fix user variable handling and preview

stefansjfw committed:
Merge remote-tracking branch 'origin/main' into stefan/env_var_new

stefansjfw committed:
spellcheck

stefansjfw committed:
Try fix spellcheck

stefansjfw committed:
Revert "spellcheck"

This reverts commit ee76231974ab6bee8d788bfa993f2daf5ba23e73.

stefansjfw committed:
Revert "Try fix spellcheck"

This reverts commit dc8f04afb976aa137e504934d263988676f7c633.

stefansjfw committed:
Merge remote-tracking branch 'origin/main' into stefan/env_var_new

stefansjfw committed:
Fix path and duplicates conflict

stefansjfw committed:
Fix PATH handling
Fix unapply on delete active variable
Fix ordering in applied variables

stefansjfw commented:
/azp run

stefansjfw commented:
/azp run

stefansjfw committed:
Merge remote-tracking branch 'origin/main' into stefan/env_var_new

stefansjfw committed:
Show variables as lists and add drag-to-reorder feature

htcfreek commented:
> "List" style vars

How do we detect that? The name of PATH var or the `;`? Because a `;` is not always an indication for a "List" style var.

stefansjfw commented:
> > "List" style vars
> 
> How do we detect that? The name of PATH var or the `;`? Because a `;` is not always an indication for a "List" style var.

The name it is

stefansjfw committed:
Only show specific variables as list
Update list in edit dialog on editing the value

stefansjfw committed:
spellcheck

stefansjfw committed:
Merge remote-tracking branch 'origin/main' into stefan/env_var_new

stefansjfw committed:
Update GPO policy

jaimecbernardo reviewed (changes requested):
General functionality is working well for me. Good work!
Haven't looked at the code yet.
Here's some comments from my tests:

## 1 - Deleting variables from profiles breaks expectations.
Deleting a variable from a profile is a bit weird, for example if a profile is off you can't delete it from the edit profile window unless there's a variable with the same name there already.

Also there's edit and delete buttons for profiles but only edit for the variables? You actually have to right-click to delete which feels like it's breaking expectations. Took a long time to figure that one out.
![image](https://github.com/microsoft/PowerToys/assets/26118718/1093f3b0-91de-49e7-b030-4a3e906bdb7d)

## 2 - System-wide variables

Looks like there's no way to apply variables to system-wide. This might be useful for Services for example, since those won't read from user-scope variables. Even if we ship without this, let's at least add a field to the variables in the profiles to differentiate them since I suspect this will be asked in the future and that way we won't need to do a save upgrade.

## 3 - Glitches when expanding default variables.

When I expand the System in Default Variables, it's quite glitchy. It flashes and sometimes disappears until I scroll down a bit and then it pops back up. Here's a screenshot of it expanded without showing anything:
![image](https://github.com/microsoft/PowerToys/assets/26118718/52c12738-94cb-4005-af98-df172a17640b)

(also nit: That blue circle likely needs an icon)

## 4 - When inside the existing variables, if we choose a variable that exists in both profiles to try to remove it's marking both system and user variables.

![image](https://github.com/microsoft/PowerToys/assets/26118718/2795f95e-a6db-4838-b2dd-863678570b01)

## 5 - Drag and drop breaks as admin

This is a known limitation for WinUI 3. You can't use drag and drop win Windows App SDK when running as admin. I suspect we might need to use move up and down buttons 🤷 .

Also drag and drop works weird in that if you drag to outside the list and then return, it seems to have lost context and the reordering freezes until you drop what you were dragging.

## 6 - Duplicated tooltip

This tooltip likely should not be a copy of the "Profiles" tooltip:
![image](https://github.com/microsoft/PowerToys/assets/26118718/cec8f139-5f99-461e-babc-78f474e9e99e)

## 7 - Editing a profile still shows as "New Profile"
![image](https://github.com/microsoft/PowerToys/assets/26118718/7a7a189e-7560-40a2-9862-3e18f0487f14)

## 8 - Backed up variables are not changed back after name change in profile variable name:
1- On your profile, create a variable that you already have already as another name and set that profile on. The backup variable will be created.
2- Change the name of the variable on your profile and disable the profile.
3 - The backup won't be reapplied.

stefansjfw committed:
Add Edit and Remove variable buttons
Remove context menu

jaimecbernardo commented on .github/actions/spell-check/expect.txt:
```diff
@@ -1932,6 +1932,7 @@ timedate
 timediff
 timeunion
 timeutil
+titlebar
```
This entry shouldn't be needed. There's an issue in some file. Let's remove it and the issue in the file too.

jaimecbernardo commented on src/modules/EnvironmentVariables/EnvironmentVariables/app.manifest:
```diff
@@ -0,0 +1,22 @@
+      <!--The ID below informs the system that this application is compatible with OS features first introduced in Windows 8. 
+      For more info see https://docs.microsoft.com/windows/win32/sysinfo/targeting-your-application-at-windows-8-1 
+      
+      It is also necessary to support features in unpackaged applications, for example the custom titlebar implementation.-->
```
```suggestion
      It is also necessary to support features in unpackaged applications, for example the custom title bar implementation.-->
```
Here's where the wrong "titlebar" entry in expect.txt came from.

jaimecbernardo reviewed (changes requested):
Adding this to my feedback, now that I looked at the code.
Great work!
It's pretty solid in my opinion. Missing things were already added to the description of this PR.
I think it could use some more telemetry events. These come to mind as useful to understand what people are using:
 - EnableProfile
 - DisableProfile
 - ChangeVariable(used the editor to change the value/name of a variable and saved the changes, perhaps a different event for each case?)
   - Profile Variable
   - User Variable
   - System Variable

stefansjfw committed:
Remove drag&drop when editing list variable and add buttons to move up/down

stefansjfw committed:
Fix Edit profile dialog title

stefansjfw committed:
Fix backup and restore variables when editing variable from applied profile

htcfreek commented:
> 2 - System-wide variables
> Looks like there's no way to apply variables to system-wide. This might be useful for Services for example, since those won't read from user-scope variables. Even if we ship without this, let's at least add a field to the variables in the profiles to differentiate them since I suspect this will be asked in the future and that way we won't need to do a save upgrade.

I think we should information bind this information to the variable to allow mixed profiles.

stefansjfw committed:
Apply var to system WIP

niels9001 committed:
Tweaks

stefansjfw committed:
Simplify edit variable logic

stefansjfw committed:
Merge branch 'stefan/env_var_new' of https://github.com/microsoft/PowerToys into stefan/env_var_new

stefansjfw committed:
Minor fixes

stefansjfw committed:
Spellcheck

stefansjfw committed:
Update src/modules/EnvironmentVariables/EnvironmentVariables/app.manifest

Co-authored-by: Jaime Bernardo <jaime@janeasystems.com>

stefansjfw committed:
spellcheck 2

stefansjfw committed:
Merge branch 'stefan/env_var_new' of https://github.com/microsoft/PowerToys into stefan/env_var_new

stefansjfw committed:
Remove unneeded string

stefansjfw commented on src/modules/EnvironmentVariables/EnvironmentVariables/Strings/en-us/Resources.resw:
```diff
@@ -0,0 +1,251 @@
+    <comment>Title of the window when running as administrator</comment>
+  </data>
+  <data name="DefaultVariables" xml:space="preserve">
+    <value>DefaultVariables</value>
```
not really.removed

stefansjfw committed:
Add more telemetry

stefansjfw committed:
Do not allow adding existing variables with the same name into the profile

stefansjfw commented:
> General functionality is working well for me. Good work! Haven't looked at the code yet. Here's some comments from my tests:
> 
> ## 1 - Deleting variables from profiles breaks expectations.
> Deleting a variable from a profile is a bit weird, for example if a profile is off you can't delete it from the edit profile window unless there's a variable with the same name there already.

Fixed

> ## 2 - System-wide variables
> Looks like there's no way to apply variables to system-wide. This might be useful for Services for example, since those won't read from user-scope variables. Even if we ship without this, let's at least add a field to the variables in the profiles to differentiate them since I suspect this will be asked in the future and that way we won't need to do a save upgrade.

TBD (maybe for v2)

> ## 3 - Glitches when expanding default variables.
> When I expand the System in Default Variables, it's quite glitchy. It flashes and sometimes disappears until I scroll down a bit and then it pops back up. Here's a screenshot of it expanded without showing anything: ![image](https://user-images.githubusercontent.com/26118718/273545520-52c12738-94cb-4005-af98-df172a17640b.png)

This is a weird one indeed. I'm investigating it.
 
> (also nit: That blue circle likely needs an icon)

Fixed
 
> ## 4 - When inside the existing variables, if we choose a variable that exists in both profiles to try to remove it's marking both system and user variables.
> ![image](https://user-images.githubusercontent.com/26118718/273546307-2795f95e-a6db-4838-b2dd-863678570b01.png)

Fixed. Adding existing variables with the same name is not allowed anymore.

> ## 5 - Drag and drop breaks as admin
> This is a known limitation for WinUI 3. You can't use drag and drop win Windows App SDK when running as admin. I suspect we might need to use move up and down buttons 🤷 .

Fixed. Added buttons to move up/down instead of drag and drop.

> ## 6 - Duplicated tooltip
> This tooltip likely should not be a copy of the "Profiles" tooltip: ![image](https://user-images.githubusercontent.com/26118718/273547823-cec8f139-5f99-461e-babc-78f474e9e99e.png)

Fixed

> ## 7 - Editing a profile still shows as "New Profile"
> ![image](https://user-images.githubusercontent.com/26118718/273547930-7a7a189e-7560-40a2-9862-3e18f0487f14.png)

Fixed
 
> ## 8 - Backed up variables are not changed back after name change in profile variable name:
> 1- On your profile, create a variable that you already have already as another name and set that profile on. The backup variable will be created. 2- Change the name of the variable on your profile and disable the profile. 3 - The backup won't be reapplied.

Fixed


Also, added suggested telemetry events.

niels9001 committed:
Adding icon

jaimecbernardo reviewed (changes requested):
LGTM! Most of my feedback looks like it's been addressed. Here's what I think it's still missing.

## 1 - Existing variables crashes after adding Path.
To reproduce:
 - Add an existing path variable from system to a profile.
 - Try to open the existing variables screen again, it crashes the application. 

## 2 - OOBE and Settings image.
I think a simple screenshot of the current state would be enough for both.

stefansjfw committed:
Fix the crash when opening existing variables dialog

stefansjfw committed:
Update Settings and OOBE screenshots

stefansjfw committed:
Fix crash when malformed profiles.json and jsonignore not needed properties

stefansjfw committed:
Fix selecting duplicates in existing variables dialog

stefansjfw commented:
> LGTM! Most of my feedback looks like it's been addressed. Here's what I think it's still missing.
> 
> ## 1 - Existing variables crashes after adding Path.
> To reproduce:
> 
> * Add an existing path variable from system to a profile.
> * Try to open the existing variables screen again, it crashes the application.
> 
> ## 2 - OOBE and Settings image.
> I think a simple screenshot of the current state would be enough for both.

Nice catch. Fixed both.

jaimecbernardo reviewed (approved):
LGTM! Great work!

htcfreek commented:
@stefansjfw
Found some UI glitches and bugs:

**1. The scroll bar on the left side overlaps the content.**
![image](https://github.com/microsoft/PowerToys/assets/61519853/4fb88c30-c599-4715-946d-abbc3fe90b70)

**2. The edit var ui is a bit confusing when editing a list variable.**
I think we should use a real grid view in that case or add a not about the use case of the list. I personally don't see any reason to use the up/down buttons as editing the text field is faster. ([This is how it looks in the Windows dialog.](https://github.com/microsoft/PowerToys/assets/61519853/47dc9a46-25e9-463d-b5fb-03e81d014ca2))
![image](https://github.com/microsoft/PowerToys/assets/61519853/0b8f24aa-77df-4afd-ad4f-f579272f50c9)

**3. There is a ui glitch for the system variables list.**
![image](https://github.com/microsoft/PowerToys/assets/61519853/61509095-0099-48b8-8d28-e6143d57cd7c)


**4. Can we change the preview of applied list variables to the same layout like in the variables lists?**
It should be a list like done on the left side of the app ui. Maybe with an expander behavior. And by the way we should add a horizontal scroll bar to the applied vars list.
![image](https://github.com/microsoft/PowerToys/assets/61519853/a7d37d13-57a8-45c9-be57-e0c22dcf5afe)

**5. Suggested improvements for the add vars content dialog when creating profiles.**
- The dialog should be centered.
- The dialog should have a cancel or close button (not using the accent color).
![image](https://github.com/microsoft/PowerToys/assets/61519853/846b3a7d-2fee-446d-afe8-09563938afb2)

htcfreek reviewed (changes requested):
@stefansjfw
Found some UI glitches and bugs:

**1. The scroll bar on the left side overlaps the content.**
![image](https://github.com/microsoft/PowerToys/assets/61519853/4fb88c30-c599-4715-946d-abbc3fe90b70)

**2. The edit var ui is a bit confusing when editing a list variable.**
I think we should use a real grid view in that case or add a not about the use case of the list. I personally don't see any reason to use the up/down buttons as editing the text field is faster. ([This is how it looks in the Windows dialog.](https://github.com/microsoft/PowerToys/assets/61519853/47dc9a46-25e9-463d-b5fb-03e81d014ca2))
![image](https://github.com/microsoft/PowerToys/assets/61519853/0b8f24aa-77df-4afd-ad4f-f579272f50c9)

**3. There is a ui glitch for the system variables list.**
![image](https://github.com/microsoft/PowerToys/assets/61519853/61509095-0099-48b8-8d28-e6143d57cd7c)


**4. Can we change the preview of applied list variables to the same layout like in the variables lists?**
It should be a list like done on the left side of the app ui. Maybe with an expander behavior. And by the way we should add a horizontal scroll bar to the applied vars list.
![image](https://github.com/microsoft/PowerToys/assets/61519853/a7d37d13-57a8-45c9-be57-e0c22dcf5afe)

**5. Suggested improvements for the add vars content dialog when creating profiles.**
- The dialog should be centered.
![image](https://github.com/microsoft/PowerToys/assets/61519853/846b3a7d-2fee-446d-afe8-09563938afb2)
- The dialog should have a cancel or close button (not using the accent color).

stefansjfw committed:
Add user variable name limit (<255 chars)
Check if profile is applicable on apply
Show message if profile is not applicable

stefansjfw committed:
Merge remote-tracking branch 'origin/main' into stefan/env_var_new

stefansjfw committed:
XamlStyling

stefansjfw committed:
Better Flyout positioning
Add Cancel button to the flyout

jaimecbernardo committed:
Merge branch 'main' into pr28722

crutkas commented:
Another question:
Profiles have edit / delete inline but others need to click "..."
![image](https://github.com/microsoft/PowerToys/assets/1462282/c6b26090-de3a-49a2-941f-6c1c7f3a07ce)

crutkas commented:
With profiles, i was testing with profiles and some how managed to fubar my path with profiles
![image](https://github.com/microsoft/PowerToys/assets/1462282/edf45a91-9358-4f72-94a3-e4bfbadffbb8)

This was the settings prior.
![image](https://github.com/microsoft/PowerToys/assets/1462282/4e89404b-48ef-4f42-91b0-fcc027b4b27b)

crutkas commented:
Actually, trying to revert the change with manually typing it click and environment var wouldn't take it too :/

jaimecbernardo commented:
> i don't think the path editor is really there yet.
> 
> It feels odd to not be able to do this inline. ![image](https://user-images.githubusercontent.com/1462282/274388359-4e89404b-48ef-4f42-91b0-fcc027b4b27b.png)

We can perhaps hide the top part. 🤔

jaimecbernardo commented:
> Another question: Profiles have edit / delete inline but others need to click "..." ![image](https://user-images.githubusercontent.com/1462282/274405101-c6b26090-de3a-49a2-941f-6c1c7f3a07ce.png)

I think the idea here was not to show many inline buttons for the env vars. Right, @niels9001 ?

jaimecbernardo commented:
> With profiles, i was testing with profiles and some how managed to fubar my path with profiles ![image](https://user-images.githubusercontent.com/1462282/274408234-edf45a91-9358-4f72-94a3-e4bfbadffbb8.png)
> 
> This was the settings prior. ![image](https://user-images.githubusercontent.com/1462282/274388359-4e89404b-48ef-4f42-91b0-fcc027b4b27b.png)

Any idea how to repro it @crutkas ?

jaimecbernardo committed:
Merge branch 'main' into stefan/env_var_new

jaimecbernardo committed:
Fix UI glitches by using ItemsControl (no virtualization)

jaimecbernardo committed:
Fix spellcheck

jaimecbernardo committed:
Fix XAML style

jaimecbernardo committed:
Add horizontal scrollbar to applied variables

htcfreek commented:
> i don't think the path editor is really there yet.  
> 
> It feels odd to not be able to do this inline.
> ![image](https://github.com/microsoft/PowerToys/assets/1462282/4e89404b-48ef-4f42-91b0-fcc027b4b27b)
> 

For me the main question is if we replace the existing variable or if we add new values at the start or at the end.

jaimecbernardo commented:
The switch from ItemsRepeater to ItemsControl is causing some bugs while editing or deleting variables. I'm currently debugging, but avoid testing this PR for a while. It might delete actual user variables instead of profile variables 😬

htcfreek commented:
@jaimecbernardo
Do we backup the state before applying changes? Might be a good idea.

jaimecbernardo committed:
Revert to ItemsRepeater

jaimecbernardo commented:
> @jaimecbernardo Do we backup the state before applying changes? Might be a good idea.

No such functionality as it stands. Wonder if we should do this, since many times Env vars are used for secrets.

jaimecbernardo commented:
I've reverted to ItemsRepeater, which should fix this issue:
> It might delete actual user variables instead of profile variables 😬

It's safe to debug again while I investigate locally the difference between ItemsRepeater and ItemsControl DataContext.

jaimecbernardo committed:
Fix UI glitches by adding a decent minimum cache

jaimecbernardo commented:
I've ended up fixing the UI glitches in the expander by setting a VerticalCacheLength in a ItemsRepeater that's big enough for the items we are showing not to be recycled. Works very well for me. Just pushed.

niels9001 committed:
Fixing UI bugs

jaimecbernardo committed:
Fix spellcheck

jaimecbernardo committed:
Fix crash while trying to edit a User variable when there's no Parent
profile

jaimecbernardo committed:
Fix issue overwriting backup var when you edit var on applied profile

jaimecbernardo committed:
Fix nuking of variables when adding to applied profile

jaimecbernardo committed:
Fix profile not being saved when deleting a variable

jaimecbernardo commented:
> With profiles, i was testing with profiles and some how managed to fubar my path with profiles ![image](https://user-images.githubusercontent.com/1462282/274408234-edf45a91-9358-4f72-94a3-e4bfbadffbb8.png)
> 
> This was the settings prior. ![image](https://user-images.githubusercontent.com/1462282/274388359-4e89404b-48ef-4f42-91b0-fcc027b4b27b.png)


I've added some fixes:
- [Fix crash while trying to edit a User variable when there's no Parent](https://github.com/microsoft/PowerToys/pull/28722/commits/62ed51babfe97e75ffeb136c29fab9af86c4aefa) 
- [Fix issue overwriting backup var when you edit var on applied profile](https://github.com/microsoft/PowerToys/pull/28722/commits/5fa6d12a95645ce95ba8a6bbf89dac61991b86f2)
- [Fix nuking of variables when adding to applied profile](https://github.com/microsoft/PowerToys/pull/28722/commits/bd02b12ec5f7018d2310956ed484fb1a8b59909d)
- [Fix profile not being saved when deleting a variable](https://github.com/microsoft/PowerToys/pull/28722/commits/e33acaac1431012ab2c0741e1bae3e03227a30fe)

Two of them were ways in which the original variable could've been deleted.

jaimecbernardo committed:
Fix ValuesList empty crash, issues and no serialization

jaimecbernardo commented:
Added [Fix ValuesList empty crash, issues and no serialization](https://github.com/microsoft/PowerToys/pull/28722/commits/510024243d07f3597556ab97f7fea1a054d5629e)

It fixes some issues:
1 - An empty path variable was leading to a crash when loading it into the Edit Variable screen.
2 - A path variable with only one entry wasn't getting its view updated when editing the single entry.
3 - It was being serialized in the JSON without needed. It's now being filled after deserialization.

jaimecbernardo committed:
fix spellcheck

jaimecbernardo committed:
Allow in-line edit of list variables

jaimecbernardo commented:
I've added inline Edit of the variables that are lists.
This included changing the list ObservableCollection to have actual objects instead of direct strings, so that they can be compared by identity instead of value. Solves issues with duplicates and empty strings and whatnot.

![image](https://github.com/microsoft/PowerToys/assets/26118718/f75528ed-ef25-4afc-ac1f-4315002f59ce)

![image](https://github.com/microsoft/PowerToys/assets/26118718/9d5d8619-36d5-440d-aef4-3ce5fbc99e01)

jaimecbernardo committed:
Fix xaml style

jaimecbernardo committed:
Merge branch 'main' into pr28722

stefansjfw committed:
Merge branch 'stefan/env_var_new' of https://github.com/microsoft/PowerToys into stefan/env_var_new

stefansjfw committed:
Fix add profile variable cancel button logic

stefansjfw committed:
Fix add profile variable cancel button logic - clean the list

stefansjfw committed:
Bump VerticalCacheLength to 10

as in some cases UI glitch on expanding System variables was still present

stefansjfw committed:
Fix profile Add variable button enable/disable logic

stefansjfw committed:
Remove unneeded using

stefansjfw committed:
Merge remote-tracking branch 'origin/main' into stefan/env_var_new

stefansjfw committed:
Add to Dashboard

donlaci reviewed (approved):
LGTM, great!

stefansjfw commented:
Thank you all for your comments and contribution! Additional comments and issues can be addressed in follow-up PRs :)
